File: leetcode/go/121.best-time-to-buy-and-sell-stock.go
func maxProfit(prices []int) int {
    minp, maxp := math.MaxInt32, 0
    for _, p := range prices {
        if minp > p {
            minp = p
        } else if p - minp > maxp {
            maxp = p - minp
        }
    }
    return maxp
}

File: leetcode/rust-leetcode/src/121.best-time-to-buy-and-sell-stock.rs
impl Solution {
    pub fn max_profit(prices: Vec<i32>) -> i32 {
        let mut maxprofit:i32 = 0;
        let mut minp = prices[0];
        for p in prices {
            if p < minp {
                minp = p;
            } else if p-minp > maxprofit {
                maxprofit = p-minp;
            }
        }
        maxprofit
    }
}

/* */

use std::cmp::max;

impl Solution {
    pub fn max_profit(prices: Vec<i32>) -> i32 {
        let (mut hold, mut sell) = (i32::MIN, 0);
        for price in prices {
            sell = max(sell, hold + price);
            hold = max(hold, sell - price);
        }
        sell
    }
}


